handout/testing/services/mockbackend.md

```diff
@@ -0,0 +1,103 @@
+          connection.mockRespond(new Response(
+            new ResponseOptions({ body: mockResponse })
+          ));
+        });
+
+      searchWiki.search('Angular')
+        .subscribe(res => {
+          expect(res).toEqual(mockResponse);
+        });
+    })
+  ));
+
+});
+```
+[View Example](http://plnkr.co/edit/k5bQN2YuFMkNXrNC2Zo3?p=preview)
+
+We use `inject` to inject the `SearchWiki` service and the `MockBackend` into our test. We then wrap our entire test with a call to `fakeAsync`, which will be used to control the asynchronous behavior of the `SearchWiki` service for testing.
+
+Next, we `subscribe` to any incoming connections from our back-end. This gives us access to an object `MockConnection`, which allows us to configure the response we want to send out from our back-end, as well as test any incoming requests from the service we're testing.
+
+In our example, we want to verify that the `SearchWiki`'s search method makes a GET request to the correct URL. This is accomplished by looking at the request object we get when our `SearchWiki` service makes a connection to our mock back-end. Analyzing the `request.url` property we can see if its value is what we expect it to be. Here we are only checking the URL, but in other scenarios we can see if certain headers have been set, or if certain POST data has been sent.
+
+Now, using the `MockConnection` object we mock in some arbitrary data. We create a new `ResponseOptions` object where we can configure the properties of our response. This follows the format of a regular [Angular 2 Response class](https://angular.io/docs/js/latest/api/http/Response-class.html). Here we have simply set the `body` property to that of a basic search result set you might see from Wikipedia. We could have also set things like cookies, HTTP headers, etc., or set the `status` value to a non-200 state to test how our service responds to errors. Once we have our `ResponseOptions` configured we create a new instance of a `Respond` object and tell our back-end to start using this as a response by calling `.mockRespond`.
+
+It is possible to use multiple responses. Say your service had two possible GET requests - one for `/api/users`, and another `/api/users/1`. Each of these requests has a different corresponding set of mock data. When receiving a new connection through the `MockBackend` subscription, you can check to see what type of URL is being requested and respond with whatever set of mock data makes sense.
+
+Finally, we can test the `search` method of the `SearchWiki` service by calling it and subscribing to the result. Notice that we call `tick` before checking the contents of the search result. This is to ensure that all `Observable`s, `Promise`s, etc. and any other asynchronous operations complete before proceeding. Once our search process has finished, we check the result object to see if it contains the same data that we mocked into our back-end. If it is, then congratulations, your test has passed.
```

Comment:
Updated the plunker with another test that uses timeout. And tick needs to come before the assertion. 

```
  it('should get search results', fakeAsync(
    inject([SearchWiki], (searchWiki) => {
      searchWiki.setFoo();
      tick(1000);
      expect(searchWiki.foo).toEqual('food');
    })
  ));
```